AK/URLParser.cpp

```diff
@@ -0,0 +1,666 @@
+    auto& ascii_domain = domain;
+
+    auto forbidden_host_code_points = "\0\t\n\r #%/:<>?@[\\]^|"sv;
+    for (auto code_point : forbidden_host_code_points) {
+        if (ascii_domain.view().contains(code_point)) {
+            report_validation_error();
+            return {};
+        }
+    }
+
+    auto ipv4_host = parse_ipv4_address(ascii_domain);
+    return ipv4_host;
+}
+
+constexpr bool starts_with_windows_drive_letter(const StringView& input)
+{
+    if (input.length() < 2)
+        return false;
+    if (!is_ascii_alpha(input[0]) && !(input[1] == ':' || input[1] == '|'))
+        return false;
+    if (input.length() == 2)
+        return true;
+    return "/\\?#"sv.contains(input[2]);
+}
+
+constexpr bool is_windows_drive_letter(const StringView& input)
+{
+    return input.length() == 2 && is_ascii_alpha(input[0]) && (input[1] == ':' || input[1] == '|');
+}
+
+constexpr bool is_normalized_windows_drive_letter(const StringView& input)
+{
+    return input.length() == 2 && is_ascii_alpha(input[0]) && input[1] == ':';
+}
+
+constexpr bool is_single_dot_path_segment(const StringView& input)
+{
+    return input == "."sv || input.equals_ignoring_case("%2e"sv);
+}
+
+constexpr bool is_double_dot_path_segment(const StringView& input)
+{
+    return input == ".."sv || input.equals_ignoring_case(".%2e"sv) || input.equals_ignoring_case("%2e."sv) || input.equals_ignoring_case("%2e%2e"sv);
+}
+
+// https://fetch.spec.whatwg.org/#data-urls
+// FIXME: This only loosely follow the spec, as we use the same class for "regular" and data URLs, unlike the spec.
+Optional<URL> URLParser::parse_data_url(const String& input_string)
+{
+    dbgln_if(URL_PARSER_DEBUG, "URLParser::parse_data_url: Parsing '{}'.", input_string);
+    VERIFY(input_string.starts_with("data:"));
+    auto view = input_string.substring_view(5);
+    auto comma_offset = view.find(',');
+    if (!comma_offset.has_value())
+        return {};
+    auto mime_type = view.substring_view(0, comma_offset.value());
+    // FIXME: Strip leading and trailing ASCII whitespace from mimeType
+    auto encoded_body = view.substring_view(comma_offset.value() + 1);
+    auto body = URL::percent_decode(encoded_body);
+    bool is_base_64_encoded = false;
+    if (mime_type.ends_with(";base64", CaseSensitivity::CaseInsensitive)) {
+        is_base_64_encoded = true;
+        mime_type = mime_type.substring_view(0, mime_type.length() - 7);
+    }
+
+    if (mime_type.starts_with(";")) {
+        StringBuilder builder;
+        builder.append("text/plain");
+        builder.append(mime_type);
+        mime_type = builder.to_string();
+    }
+
+    URL url { mime_type, move(body), is_base_64_encoded };
+    dbgln_if(URL_PARSER_DEBUG, "URLParser::parse_data_url: Parsed data URL to be '{}'.", url.serialize());
+    return url;
+}
+
+// https://url.spec.whatwg.org/#concept-basic-url-parser
+// NOTE: This parser assumes a UTF-8 encoding.
+URL URLParser::parse(String input_string, Optional<URL> base_url)
```

Comment:
This was what I did at first, but this breaks the `URL(const StringView&)` constructor. I decided that if I do this, it would probably be required for everyone to use `URL::parse()` and handle the invalid URLs themselves, instead of using a constructor.
Also, `URL::is_valid` has around 30 usages all over the repository, so I decided to postpone this change. Since the URL parser is only called by the `URL` class itself, we can easily change the return type at a later stage. (This reminds me that the public APIs of `URLParser` should use `Badge<URL>`, so that nobody tries to use them directly.)

Reply:
Fair enough!